### PR TITLE
Add function to get key value from save file, called GetUserPreference.

### PR DESCRIPTION
**Pull request by SebastianMygind**

### CODE REVIEW

File: src/preferences.c [hunk 1]
```diff
@@ -77,3 +77,50 @@ void SetUserPreference(const char* key, double value) {
   fclose(preferences);
   free(kSerializedJsonNewlineWithNewline);
 }
+
+double GetUserPreference(const char* key) {
+  double value = 0;
+  FILE* user_preferences = fopen("preferences.json", "r");
+
+  if (user_preferences == NULL) {
```
Comment (audio-engineer) [inline src/preferences.c]:
```suggestion
  if (!user_preferences) {
```

File: src/application.c [hunk 2]
```diff
@@ -23,7 +23,6 @@ void Run() {
   SetUserPreference("environment", 0.30);
   SetUserPreference("health", 0.40);
   // NOLINTEND(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
-
```
Comment (audio-engineer) [inline src/application.c]: @SebastianMygind Why?

File: src/preferences.c [hunk 3]
```diff
@@ -77,3 +77,50 @@ void SetUserPreference(const char* key, double value) {
   fclose(preferences);
   free(kSerializedJsonNewlineWithNewline);
 }
+
+double GetUserPreference(const char* key) {
```
Comment (audio-engineer) [inline src/preferences.c]:
```suggestion
double GetUserPreference(const char* const key) {
```

File: src/preferences.c [hunk 4]
```diff
@@ -77,3 +77,50 @@ void SetUserPreference(const char* key, double value) {
...
+  if (user_preferences == NULL) {
+    printf("Error opening user preferences file!\n");
+    return value;
+  }
+
+  fseek(user_preferences, 0, SEEK_END);
+  unsigned long kFileSize = (unsigned long)ftell(user_preferences);
+  fseek(user_preferences, 0, SEEK_SET);
+
+  char* kFileBuffer = (char*)calloc(kFileSize + 1, sizeof(char));
```
Comment (audio-engineer) [inline src/preferences.c]:
```suggestion
  char* kFileBuffer = calloc(kFileSize + 1, sizeof(char));
```

File: src/preferences.c [hunk 5]
```diff
@@ -77,3 +77,50 @@ void SetUserPreference(const char* key, double value) {
...
+  if (kFileBuffer == NULL) {
```
Comment (audio-engineer) [inline src/preferences.c]:
```suggestion
  if (!kFileBuffer) {
```

File: src/preferences.c [hunk 6]
```diff
@@ -77,3 +77,50 @@ void SetUserPreference(const char* key, double value) {
...
+    printf("Error allocating memory for file buffer!\n");
+    fclose(user_preferences);
+    return value;
+  }
+  size_t bytes_read =
```
Comment (audio-engineer) [inline src/preferences.c]:
```suggestion
  const size_t kBytes =
```

File: src/preferences.c [hunk 7]
```diff
@@ -77,3 +77,50 @@ void SetUserPreference(const char* key, double value) {
...
+      fread(kFileBuffer, sizeof(char), kFileSize, user_preferences);
+  fclose(user_preferences);
+
+  if (bytes_read != (size_t)kFileSize) {
```
Comment (audio-engineer) [inline src/preferences.c]: Ok, but why cast to `size_t`?

File: src/preferences.c [hunk 8]
```diff
@@ -77,3 +77,50 @@ void SetUserPreference(const char* key, double value) {
...
+    printf("Error reading user preferences file!\n");
+    free(kFileBuffer);
+    return value;
+  }
+
+  // kFileBuffer contains the content of the file
+  cJSON* save_file = cJSON_Parse(kFileBuffer);
+  free(kFileBuffer);
+
+  if (save_file == NULL) {
```
Comment (audio-engineer) [inline src/preferences.c]:
```suggestion
  if (!json_preferences) {
```

File: src/preferences.c [hunk 9]
```diff
@@ -77,3 +77,50 @@ void SetUserPreference(const char* key, double value) {
...
+  fclose(user_preferences);
+
+  if (bytes_read != (size_t)kFileSize) {
+    printf("Error reading user preferences file!\n");
+    free(kFileBuffer);
+    return value;
+  }
+
+  // kFileBuffer contains the content of the file
+  cJSON* save_file = cJSON_Parse(kFileBuffer);
```
Comment (audio-engineer) [inline src/preferences.c]:
```suggestion
  cJSON* json_preferences = cJSON_Parse(kFileBuffer);
```
Maybe?

File: src/preferences.c [hunk 10]
```diff
@@ -77,3 +77,50 @@ void SetUserPreference(const char* key, double value) {
...
+  if (save_file == NULL) {
+    printf("Error parsing JSON file!\n");
+    return value;
+  }
+
+  cJSON* file_item = cJSON_GetObjectItem(save_file, key);
+
+  value = cJSON_GetNumberValue(file_item);
+
+  cJSON_Delete(save_file);
```
Comment (audio-engineer) [inline src/preferences.c]:
```suggestion
  cJSON_Delete(json_preferences);
```

**Review by audio-engineer — CHANGES_REQUESTED** [review]
Good job 👍

**Review by MadsHeilmann — APPROVED** [review]

File: src/application.c (hunk 2 above)
Comment (SebastianMygind) [reply src/application.c]: a mistake. If i am understanding correctly that I removed a newline

File: src/preferences.c (hunk 7 above)
Comment (SebastianMygind) [reply src/preferences.c]: Copy pasted... Shoud it be changed?

Comment (audio-engineer) [reply src/preferences.c]: Redundant casts are... redundant. Try to remove it and see if CLion complains, i.e. suggests another cast. If not, you are good to go

Comment (audio-engineer) [reply src/preferences.c]: It seems like `unsigned long` and `size_t` are comparable:
image

**Review by TexWO — APPROVED** [review]

File: src/application.c (hunk 2 above)
Comment (audio-engineer) [reply src/application.c]: Yeah, you did, please bring it back

**Review by audio-engineer — CHANGES_REQUESTED** [review]

**Review by audio-engineer — APPROVED** [review]

**Pull request merged**